flink-kubernetes/src/main/java/org/apache/flink/kubernetes/KubernetesResourceManagerDriver.java

```diff
@@ -0,0 +1,325 @@
+
+	public KubernetesResourceManagerDriver(
+			Configuration flinkConfig,
+			FlinkKubeClient kubeClient,
+			KubernetesResourceManagerDriverConfiguration configuration) {
+		super(flinkConfig, GlobalConfiguration.loadConfiguration());
+
+		this.clusterId = Preconditions.checkNotNull(configuration.getClusterId());
+		this.podCreationRetryInterval = Preconditions.checkNotNull(configuration.getPodCreationRetryInterval());
+		this.kubeClient = Preconditions.checkNotNull(kubeClient);
+		this.requestResourceFutures = new HashMap<>();
+		this.podCreationCoolDown = FutureUtils.completedVoidFuture();
+	}
+
+	// ------------------------------------------------------------------------
+	//  ResourceManagerDriver
+	// ------------------------------------------------------------------------
+
+	@Override
+	protected void initializeInternal() throws Exception {
+		recoverWorkerNodesFromPreviousAttempts();
+
+		podsWatch = kubeClient.watchPodsAndDoCallback(
+				KubernetesUtils.getTaskManagerLabels(clusterId),
+				this);
+	}
+
+	@Override
+	public CompletableFuture<Void> terminate() {
+		// shut down all components
+		Exception exception = null;
+
+		try {
+			podsWatch.close();
+		} catch (Exception e) {
+			exception = e;
+		}
+
+		try {
+			kubeClient.close();
+		} catch (Exception e) {
+			exception = ExceptionUtils.firstOrSuppressed(e, exception);
+		}
+
+		return exception == null ?
+				FutureUtils.completedVoidFuture() :
+				FutureUtils.completedExceptionally(exception);
+	}
+
+	@Override
+	public void deregisterApplication(ApplicationStatus finalStatus, @Nullable String optionalDiagnostics) {
+		log.info("Deregistering flink kubernetes cluster, clusterId: {}, diagnostics: {}",
+				clusterId,
+				optionalDiagnostics == null ? "" : optionalDiagnostics);
+		kubeClient.stopAndCleanupCluster(clusterId);
+	}
+
+	@Override
+	public CompletableFuture<KubernetesWorkerNode> requestResource(TaskExecutorProcessSpec taskExecutorProcessSpec) {
+		final KubernetesTaskManagerParameters parameters =
+				createKubernetesTaskManagerParameters(taskExecutorProcessSpec);
+		final KubernetesPod taskManagerPod =
+				KubernetesTaskManagerFactory.buildTaskManagerKubernetesPod(parameters);
+		final String podName = taskManagerPod.getName();
+		final CompletableFuture<KubernetesWorkerNode> requestResourceFuture = new CompletableFuture<>();
+
+		requestResourceFutures.put(podName, requestResourceFuture);
+
+		log.info("Creating new TaskManager pod with name {} and resource <{},{}>.",
+				podName,
+				parameters.getTaskManagerMemoryMB(),
+				parameters.getTaskManagerCPU());
+
+		// When K8s API Server is temporary unavailable, `kubeClient.createTaskManagerPod` might fail immediately.
+		// In case of pod creation failures, we should wait for an interval before trying to create new pods.
+		// Otherwise, ActiveResourceManager will always re-requesting the worker, which keeps the main thread busy.
+		final CompletableFuture<Void> createPodFuture =
```

Comment:
We have behavior change about retrying interval here. But i think is reasonable and do not increase the pressure of K8s apiserver.

Reply:
True, there is indeed a behavior change.
Unfortunately, I don't find good way to not introducing such behavior changes with the refactoring, since the re-request logic has been abstracted to the `ActiveResourceManager`.
As far as I can see, this behavior change should not introduce any regression.

Reply:
We might think about whether such a mechanism would make sense for all `ActiveResourceManagers`. If the answer is yes, then we could move the retry logic into the `ActiveResourceManager` class.

Reply:
My answer would be no.

On Yarn, when calling `AMRMClientAsync.addContainerRequest`, the request is first added to the local client, then sent to Yarn RM via heartbeats. This prevents the requests from being failed and re-requested immediately.